platforms/nuttx/src/px4/common/px4_mtd.cpp

```diff
@@ -0,0 +1,458 @@
+	.devid    = SPIDEV_FLASH(0)
+};
+
+const px4_mtd_entry_t fram = {
+	.device = &spifram,
+	.npart = 2,
+	.partd = {
+		{
+			.type = MTD_PARAMETERS,
+			.path = "/fs/mtd_params",
+			.nblocks = 32
+		},
+		{
+			.type = MTD_WAYPOINTS,
+			.path = "/fs/mtd_waypoints",
+			.nblocks = 32
+
+		}
+	},
+};
+
+
+static const px4_mtd_manifest_t default_mtd_config = {
+	.nconfigs   = 1,
+	.entries = {
+		&fram,
+	}
+};
+#endif
+
+int px4_mtd_config(const px4_mtd_manifest_t *mft_mtd)
+{
+	int rv = -EINVAL;
+
+	const px4_mtd_manifest_t *mtd_list = mft_mtd ? mft_mtd : &default_mtd_config;
+
+	if (mtd_list == nullptr) {
+		PX4_ERR("Invalid mtd configuration!");
+		return rv;
+	}
+
+	if (mtd_list->nconfigs == 0) {
+		return 0;
+	}
+
+	rv = -ENOMEM;
+	int total_blocks = 0;
+
+	instances = new mtd_instance_s[mtd_list->nconfigs];
+
+	if (instances == nullptr) {
+memoryout:
+		PX4_ERR("failed to allocate memory!");
+		return rv;
+	}
+
+	for (uint32_t i = 0; i < mtd_list->nconfigs; i++) {
+		num_instances++;
+		uint32_t nparts = mtd_list->entries[i]->npart;
+		instances[i].devid = mtd_list->entries[i]->device->devid;
+		instances[i].mtd_dev = nullptr;
+
+		rv = -ENOMEM;
+		instances[i].part_dev = new FAR struct mtd_dev_s *[nparts];
+
+		if (instances[i].part_dev == nullptr) {
+			goto memoryout;
+		}
+
+		instances[i].partition_block_counts = new int[nparts];
+
+		if (instances[i].partition_block_counts == nullptr) {
+			goto memoryout;
+		}
+
+		instances[i].partition_types = new int[nparts];
+
+		if (instances[i].partition_types == nullptr) {
+			goto memoryout;
+		}
+
+		instances[i].partition_names = new const char *[nparts];
+
+		if (instances[i].partition_names == nullptr) {
+			goto memoryout;
+		}
+
+		for (uint32_t p = 0; p < nparts; p++) {
+			instances[i].partition_block_counts[p] =  mtd_list->entries[i]->partd[p].nblocks;
+			instances[i].partition_names[p] = mtd_list->entries[i]->partd[p].path;
+			instances[i].partition_types[p] = mtd_list->entries[i]->partd[p].type;
+		}
+
+		instances[i].n_partitions_current = nparts;
+
+		if (mtd_list->entries[i]->device->bus_type == px4_mft_device_t::I2C) {
+			rv = at24xxx_attach(instances[i]);
+
+		} else {
+			rv = ramtron_attach(instances[i]);
+		}
+
+		if (rv != 0) {
+			goto errout;
+		}
+
+		unsigned long blocksize;
+		unsigned long erasesize;
+		unsigned long neraseblocks;
+		unsigned int  blkpererase;
+		unsigned int  nblocks;
+		unsigned int  partsize;
+
+		rv = px4_mtd_get_geometry(&instances[i], &blocksize, &erasesize, &neraseblocks, &blkpererase, &nblocks, &partsize);
+
+		if (rv != 0) {
+			goto errout;
+		}
+
+		/* Now create MTD FLASH partitions */
+
+		char blockname[32];
+
+		unsigned offset;
+		unsigned part;
+
+		for (offset = 0, part = 0; rv == 0 && part < nparts; offset += instances[i].partition_block_counts[part], part++) {
+
+			/* Create the partition */
+
+			instances[i].part_dev[part] = mtd_partition(instances[i].mtd_dev, offset, instances[i].partition_block_counts[part]);
```

Comment:
I didn't test this PR, but from code it looks like that here you still have the same mistake I mentioned https://github.com/PX4/PX4-Autopilot/pull/15966#pullrequestreview-535777030: NuttX mtd_partition [expects](https://github.com/PX4/NuttX/blob/904a602c74dc08a100b5c2bd490807de19e73e10/drivers/mtd/mtd_partition.c#L745-L748) the number of blocks while `instances[i].partition_block_counts[part]` you are setting and use is the number of **erase blocks**, so you partition in fact will be blkpererase times less (number of blocks in erase block)!

In addition you also can confirm correct using of `mtd_partition()` in the NuttX [mtdpart app example](https://github.com/PX4/NuttX-apps/blob/909eff540d8004371d5838b9c0764cc753d665cf/examples/mtdpart/mtdpart_main.c#L228-L230).

As I understand you are going to set number of erase blocks in manifests, I suggest renaming it and fix partition creation something like this:
`instances[i].part_dev[part] = mtd_partition(instances[i].mtd_dev, offset, instances[i].partition_eraseblock_counts[part]*blkpererase);`